pallets/subtensor/tests/epoch.rs

```diff
@@ -583,6 +583,298 @@ fn test_16384_graph_sparse() {
 	});
 }
 
+// Test bonds exponential moving average over a sequence of epochs.
+#[test]
+fn test_bonds() {
+	new_test_ext().execute_with(|| {
+		let sparse: bool = true;
+		let n: u16 = 8;
+		let netuid: u16 = 0;
+		let tempo: u16 = u16::MAX - 1;  // high tempo to skip automatic epochs in on_initialize, use manual epochs instead
+		let block_number: u64 = 0;
+		let max_stake: u64 = 4;
+		let stakes: Vec<u64> = vec![1, 2, 3, 4, 0, 0, 0, 0];
+		add_network(netuid, tempo, 0);
+		SubtensorModule::set_max_allowed_uids( netuid, n );
+		assert_eq!(SubtensorModule::get_max_allowed_uids(netuid), n);
+		SubtensorModule::set_max_registrations_per_block( netuid, n );
+		SubtensorModule::set_target_registrations_per_interval(netuid, n);
+
+		// === Register [validator1, validator2, validator3, validator4, server1, server2, server3, server4]
+		for key in 0..n as u64 {
+			SubtensorModule::add_balance_to_coldkey_account( &U256::from(key), max_stake );
+			let (nonce, work): (u64, Vec<u8>) = SubtensorModule::create_work_for_block_number( netuid, block_number, key * 1_000_000, &U256::from(key));
+			assert_ok!(SubtensorModule::register(<<Test as Config>::RuntimeOrigin>::signed(U256::from(key)), netuid, block_number, nonce, work, U256::from(key), U256::from(key)));
+			SubtensorModule::increase_stake_on_coldkey_hotkey_account( &U256::from(key), &U256::from(key), stakes[key as usize] );
+		}
+		assert_eq!(SubtensorModule::get_max_allowed_uids(netuid), n);
+		assert_eq!(SubtensorModule::get_subnetwork_n(netuid), n);
+
+		// === Issue validator permits
+		assert_ok!( SubtensorModule::sudo_set_max_allowed_validators(<<Test as Config>::RuntimeOrigin>::root(), netuid, n) );
+		assert_eq!( SubtensorModule::get_max_allowed_validators(netuid), n);
+		SubtensorModule::epoch( netuid, 1_000_000_000 ); // run first epoch to set allowed validators
+		run_to_block( 1 ); // run to next block to ensure weights are set on nodes after their registration block
+
+		// === Set weights [val->srv1: 0.1, val->srv2: 0.2, val->srv3: 0.3, val->srv4: 0.4]
+		for uid in 0..(n/2) as u64 {
+			assert_ok!(SubtensorModule::set_weights(RuntimeOrigin::signed(U256::from(uid)), netuid, ((n/2)..n).collect(), vec![ u16::MAX/4, u16::MAX/2, (u16::MAX/4)*3, u16::MAX], 0));
+		}
+		if sparse { SubtensorModule::epoch( netuid, 1_000_000_000 ); }
+		else { SubtensorModule::epoch_dense( netuid, 1_000_000_000 ); }
+		/*  n: 8
```

Comment:
:question: Any way to have these new commented bits be asserted as true by tests?

Reply:
It is possible, but the focus with `test_bonds` is just to check the bond carry-over across epochs, so the limited asserts suffice.